bluepyopt/ephys/parameters.py

```diff
@@ -19,95 +19,47 @@
  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 """
 
+from abc import abstractmethod
+
 import logging
 
+import bluepyopt
 from . import parameterscalers
 
 logger = logging.getLogger(__name__)
 
 # TODO location and stimulus parameters should also be optimisable
 
 
-class Parameter(object):
-
-    """Base parameter class"""
-
-    def __init__(self, name, value=None, frozen=False, bounds=None):
-        """Constructor"""
-
-        if frozen and value is None:
-            raise Exception('Must provide a value for frozen parameters')
-
-        self.name = name
-        self.prefix = None
-        self.bounds = bounds
-        self._value = value
-        self.check_bounds()
-        self.frozen = frozen
-
-    @property
-    def lower_bound(self):
-        """Lower bound"""
-        if self.bounds is not None:
-            return self.bounds[0]
-        else:
-            return None
-
-    @property
-    def upper_bound(self):
-        """Lower bound"""
-        if self.bounds is not None:
-            return self.bounds[1]
-        else:
-            return None
-
-    @property
-    def value(self):
-        """Parameter value"""
-        return self._value
-
-    def freeze(self, value):
-        """Freeze parameter to certain value"""
-        self.value = value
-        self.frozen = True
-
-    def unfreeze(self, value=None):
-        """Unfreeze parameter, if desired, set it to a certain value"""
-        self._value = value
-        self.frozen = False
-
-    @value.setter
-    def value(self, value):
-        """Set parameter value"""
-        if self.frozen:
-            raise Exception(
-                'Parameter: parameter %s is frozen, unable to change value' %
-                self.name)
-        else:
-            self._value = value
-            self.check_bounds()
-
-    def check_bounds(self):
-        """Check if parameter is within bounds"""
-        if self.bounds and self._value is not None:
-            if not self.lower_bound <= self._value <= self.upper_bound:
-                raise Exception(
-                    'Parameter %s has value %s outside of bounds [%s, %s]' %
-                    (self.name, self._value, str(self.lower_bound),
-                     str(self.upper_bound)))
+class NrnParameter(bluepyopt.parameters.Parameter):
 
-    def destroy(self):
-        """Destroy parameter instantation"""
+    """Abstract Parameter class for Neuron object parameters"""
+
+    def __init__(
+            self,
+            name,
+            value=None,
+            frozen=False,
+            bounds=None):
+        """Contructor"""
+
+        super(NrnParameter, self).__init__(
+            name,
+            value=value,
+            frozen=frozen,
+            bounds=bounds)
+
+    @abstractmethod
+    def instantiate(self, sim=None, icell=None):
+        """Instantiate the parameter in the simulator"""
         pass
 
-    def raise_if_no_value(self):
```

Comment:
But it also means one has to write more tests across this error path, and when there's an update, it has to be applied in multiple places.